File: 0921-spiral-matrix-iii/0921-spiral-matrix-iii.py
class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> list[list[int]]:
        matrix = [[0, 0] for _ in range(rows * cols)]
        k = 0
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        dir = 0
        steps = 1

        while k < len(matrix):
            for i in range(steps):
                if 0 <= rStart < rows and 0 <= cStart < cols:
                    matrix[k] = [rStart, cStart]
                    k += 1
                rStart += directions[dir][0]
                cStart += directions[dir][1]
            dir = (dir + 1) % 4
            if dir % 2 == 0:
                steps += 1

        return matrix

File: 0921-spiral-matrix-iii/0921-spiral-matrix-iii.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        vector<vector<int>> matrix(rows * cols, vector<int>(2));
        int k = 0;
        int directions[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int dir = 0;
        int steps = 1;

        while (k < matrix.size()) {
            for (int i = 0; i < steps; i++) {
                if (rStart >= 0 && rStart < rows && cStart >= 0 && cStart < cols) {
                    matrix[k][0] = rStart;
                    matrix[k][1] = cStart;
                    k++;
                }
                rStart += directions[dir][0];
                cStart += directions[dir][1];
            }
            dir = (dir + 1) % 4;
            if (dir % 2 == 0) {
                steps++;
            }
        }
        return matrix;
    }
};